### PR TITLE
CORS Middleware

### PR DESCRIPTION
**Pull request by dranikpg**
Hi! I've implemented middleware for supporting CORS headers like discussed [here](https://github.com/CrowCpp/Crow/issues/224).

### What it looks like

CORSHandler is responsible for setting headers in after_handle. By default, it already enables CORS after being included as a middleware. The global rules and rules for certain prefixes can be changed as follows:

```c++
    crow::App<crow::CORSHandler> app;

    auto& cors = app.get_middleware<crow::CORSHandler>();

    cors
      .global()
        .methods("POST"_method, "GET"_method)
      .prefix("/cors")
        .origin("example.com")
        .headers("X-Custom-Header")
      .blueprint(bp)
        .ignore();
```

It resembles [Flasks resource specific CORS](https://flask-cors.readthedocs.io/en/latest/#resource-specific-cors) to some extent. But with a builder instead of dicts 🙂 

### What I've changed

`middlewares/cors.h` now contains the CORSHandler with CORSRules.

`examples/middlewares/cors.h` has a simple example for setting cors rules.

I've also added a simple test.

### Issues

1. Middleware is initialized by the app. This means that an additional 
```c++
auto& cors = app.get_middleware<crow::CORSHandler>();
```
is required for setting specific rules.

I also though of implementing a custom cors handler, which extends the default one

```c++
class MyCORS : public CORSHandler 
{
  MyCORS() 
  {
    // customize here
  }
}
``` 
This makes the code less imperative, but it makes it more difficult to customize it's behaviour dynamically (i.e. use variables from the main scope). This would also allow overwriting rules for handlers with local middleware.

2. Only one origin can be specified (The origin header can return only one). But at least some web frameworks allow specifying multiple, from which the one is automatically selected (if its the same origin)

3. There is no inheritance from global to prefix rules, i.e. rules are applied either from a prefix **or** the global handler. Should global rules be applied always? Or should they be independent from prefix rules?

### CODE REVIEW

File: include/crow/middlewares/cors.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+        // Set response headers
+        void apply(crow::response& res)
+        {
+            if (ignore_) return;
+            set_header("Access-Control-Allow-Origin", origin_, res);
+            set_header("Access-Control-Allow-Methods", methods_, res);
+            set_header("Access-Control-Allow-Headers", headers_, res);
+            set_header("Access-Control-Max-Age", max_age_, res);
+            if (allow_credentials_) set_header("Access-Control-Allow-Credentials", "true", res);
```
Comment (dranikpg) [inline include/crow/middlewares/cors.h:98]: Those are the most common CORS headers used (at least the first three).

File: include/crow/middlewares/cors.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        {}
+
+        void after_handle(crow::request& req, crow::response& res, context& /*ctx*/)
+        {
+            auto& rule = find_rule(req.url);
+            rule.apply(res);
+        }
+
+        // Handle CORS on specific prefix path
+        CORSRules& prefix(const std::string& prefix)
```
Comment (dranikpg) [inline include/crow/middlewares/cors.h:130]: I though of using a more flexible builder like
```c++
cors
  .prefix("/foo")
    .origin()
    .header()
  .prefix("/bar")
    .origin()
    .header()
```
but clang-format doesn't like those and aligns all calls 😢 What do you think of those?Seems to be convenient

File: include/crow/middlewares/cors.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        // Global CORS policy
+        CORSRules& global()
+        {
+            return default_;
+        }
+
+    private:
+        CORSRules& find_rule(const std::string& path)
+        {
+            for (auto& rule : rules)
```
Comment (dranikpg) [inline include/crow/middlewares/cors.h:145]: I expect that there won't be more than a couple of prefixes, so linear search is okay. Something like a trie (there is no generic one in CROW) is a definite overkill.

File: include/crow/middlewares/cors.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            {
+                if (path.rfind(rule.first, 0) == 0)
```
Comment (dranikpg) [inline include/crow/middlewares/cors.h:147]: that is STL's way of string.startsWith() 😆  Maybe I should comment this

File: include/crow/middlewares/cors.h [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    // CORSRules is used for tuning cors policies
+    struct CORSRules
+    {
+        friend struct crow::CORSHandler;
+
+        // Set Access-Control-Allow-Origin. Default is "*"
+        CORSRules& origin(const std::string& origin)
+        {
+            origin_ = origin;
```
Comment (dranikpg) [inline include/crow/middlewares/cors.h:17]: We could elide copies here by making the code more complex. Given that this is called once on startup, I don't think its worth it

File: include/crow/middlewares/cors.h [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            if (list == "*") list = "";
+            if (list.size() > 0) list += ", ";
+            list += val;
+        }
+
+        // Set header `key` to `value` if it is not set
+        void set_header(const std::string& key, const std::string& value, crow::response& res)
+        {
+            if (value.size() == 0) return;
+            if (!get_header_value(res.headers, key).empty()) return;
```
Comment (dranikpg) [inline include/crow/middlewares/cors.h:86]: Headers are **not** overwritten, this means that a handler can change the CORS behaviour if it wants to

**Review by The-EDev — CHANGES_REQUESTED** [review]

**Review by The-EDev — APPROVED** [review]
Seems all good, merging now.

**Pull request merged**